src/Microsoft.Diagnostics.Monitoring/ReversedServerConnectionsSource.cs

```diff
@@ -0,0 +1,194 @@
+                    .Select(c => new DiagnosticsConnection(c))
+                    .ToList()
+                    .AsReadOnly();
+            }
+            finally
+            {
+                _connectionsSemaphore.Release();
+            }
+        }
+
+        /// <summary>
+        /// Accepts connections from the reversed diagnostics server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        private async Task AcceptAsync(CancellationToken token)
+        {
+            // Continuously accept connections from the reversed diagnostics server so
+            // that <see cref="ReversedDiagnosticsServer.AcceptAsync(CancellationToken)"/>
+            // is always awaited in order to to handle new runtime instance connections
+            // as well as existing runtime instance reconnections.
+            while (!token.IsCancellationRequested)
+            {
+                try
+                {
+                    ReversedDiagnosticsConnection connection = await _server.AcceptAsync(token);
+
+                    await _connectionsSemaphore.WaitAsync(token);
+
+                    _connections.Add(connection);
+
+                    _connectionsSemaphore.Release();
+
+                    if (null != _newConnectionSource)
+                    {
+                        _newConnectionSource.TrySetResult(null);
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                }
+            }
+        }
+
+        /// <summary>
+        /// Tests a diagnostics connection and removes it if the associated runtime instance is no longer active.
+        /// </summary>
+        private async Task PruneConnectionAsync(ReversedDiagnosticsConnection connection, CancellationToken token)
+        {
+            // Cancel event pipe processing within reasonable amount of time to remove potentially unresponsive processes.
+            using CancellationTokenSource timeoutSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+
+            // Try to get ProcessInfo information to test the connection
+            try
+            {
+                DiagnosticsClient client = new DiagnosticsClient(connection.Endpoint);
+
+                // NOTE: This test does not need to necessarily use the event pipe providers and collection.
+                // It only needs to send some type of request and receive a response from the runtime instance
+                // to verify that the diagnostics pipe/socket is viable and the runtime instance is still connected.
+                await using DiagnosticsEventPipeProcessor process = new DiagnosticsEventPipeProcessor(PipeMode.ProcessInfo);
+
+                using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, timeoutSource.Token);
+
+                await process.Process(client, 0, Timeout.InfiniteTimeSpan, linkedSource.Token);
```

Comment:
Consider a non-infinite timespan, and if it can't be done in time, we consider it invalid.